File: radar_lassez/nodes/publisher.ts
import { payloadClient } from '../lib/payload-client';
import { getEffectiveParam } from '../lib/config-resolver';
import { TwitterApi } from 'twitter-api-v2';
import { BskyAgent } from '@atproto/api';

/**
 * Convertit une couleur au format Hexadécimal (#DC2626) en valeur entière pour Discord
 */
function hexToDecimal(hex: string): number {
    const cleanedHex = hex.replace('#', '');
    return parseInt(cleanedHex, 16);
}

/**
 * Génère un entier aléatoire dans une plage [min, max] pour espacer les publications
 */
function getRandomInt(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

/**
 * Nœud 6 : Tour de Contrôle (Publisher / Diffusion & Réseaux Sociaux)
 *
 * Orchestre en 2 phases distinctes :
 * Phase A (L'Enfileur) : Récupère les articles qualifiés (PENDING) et programme des missions
 *                         pour chaque réseau social (Discord, X/Twitter, Bluesky, Mastodon, Payload CMS).
 * Phase B (Le Diffuseur) : Déclenche les envois réels prévus à l'heure courante via les APIs tierces.
 */
export async function runPublisherNode() {
    console.log(`\n🚀 [Node 6: Publisher] Lancement de la Tour de Contrôle (Scheduling & Diffusion)`);

    // Chargement des paramètres de diffusion
    const enableDiscord = await getEffectiveParam('publisher', 'enableDiscord', true);
    const enableX = await getEffectiveParam('publisher', 'enableX', false);
    const enableMastodon = await getEffectiveParam('publisher', 'enableMastodon', false);
    const enableBluesky = await getEffectiveParam('publisher', 'enableBluesky', false);
    const enablePayloadCMS = await getEffectiveParam('publisher', 'enablePayloadCMS', true);
    
    const minDelay = await getEffectiveParam('publisher', 'minPublishDelay', 60);
    const maxDelay = await getEffectiveParam('publisher', 'maxPublishDelay', 120);
    const enableAutoPublish = await getEffectiveParam('publisher', 'enableAutoPublish', true);

    // ========================================================
    // PHASE A : CRÉATION DES MISSIONS DE PUBLICATION
    // ========================================================
    console.log(`🚀 [Node 6: Phase A] Recherche de nouveaux articles à planifier...`);

    const pendingTopics = await payloadClient.getPendingSignalsWithoutPublications();

    if (pendingTopics.length > 0) {
        console.log(`🚀 [Node 6: Phase A] 📤 ${pendingTopics.length} nouveaux articles à programmer.`);

        // Bascule atomique PENDING → QUEUED (les topics récupérés n'ont pas de publication)
        await payloadClient.updateManySignals(
            pendingTopics.map(t => t.id),
            { status: 'QUEUED' }
        );

        const platforms = [];
        if (enableDiscord) platforms.push({ name: 'DISCORD', mode: await getEffectiveParam('publisher', 'discordPublishMode', 'DIRECT') });
        if (enableX) platforms.push({ name: 'X', mode: await getEffectiveParam('publisher', 'xPublishMode', 'SCHEDULED') });
        if (enableBluesky) platforms.push({ name: 'BLUESKY', mode: await getEffectiveParam('publisher', 'blueskyPublishMode', 'SCHEDULED') });
        if (enableMastodon) platforms.push({ name: 'MASTODON', mode: await getEffectiveParam('publisher', 'mastodonPublishMode', 'SCHEDULED') });
        if (enablePayloadCMS) platforms.push({ name: 'PAYLOAD', mode: await getEffectiveParam('publisher', 'payloadPublishMode', 'DIRECT') });

        const lastScheduledDates: Record<string, Date> = {};
        for (const p of platforms) {
            if (p.mode === 'SCHEDULED') {
                const lastPub = await payloadClient.getLastScheduledPublication(p.name);
                lastScheduledDates[p.name] = lastPub?.scheduledAt && lastPub.scheduledAt > new Date() ? lastPub.scheduledAt : new Date();
            } else {
                lastScheduledDates[p.name] = new Date();
            }
        }

        const publicationsToCreate = [];

        for (const topic of pendingTopics) {
            for (const platform of platforms) {
                let finalScheduledAt = new Date();

                if (platform.mode === "SCHEDULED") {
                    let baseDate = lastScheduledDates[platform.name];
                    const delayMinutes = getRandomInt(Number(minDelay), Number(maxDelay));
                    finalScheduledAt = new Date(baseDate.getTime() + delayMinutes * 60000);
                    lastScheduledDates[platform.name] = finalScheduledAt;
                }

                publicationsToCreate.push({
                    topicId: topic.id,
                    platform: platform.name,
                    status: 'PENDING',
                    scheduledAt: finalScheduledAt
                });
            }
        }

        if (publicationsToCreate.length > 0) {
            await payloadClient.createPublications(publicationsToCreate);
            console.log(`🚀 [Node 6: Phase A] ✅ ${publicationsToCreate.length} missions créées dans Payload.`);
        }
    }

    // ========================================================
    // PHASE B : EXÉCUTION DES MISSIONS PRÊTES A PARTIR
    // ========================================================
    console.log(`🚀 [Node 6: Phase B] Vérification des publications programmées prêtes...`);

    if (!enableAutoPublish) {
        console.log(`🚀 [Node 6: Phase B] ⏸️ Pilote automatique désactivé (enableAutoPublish=false). Diffusion ignorée.`);
        return;
    }

    const duePublications = await payloadClient.getDuePublications(10);

    if (duePublications.length === 0) {
        console.log(`🚀 [Node 6: Phase B] 📭 Aucune publication en attente pour l'instant.`);
        return;
    }

    console.log(`🚀 [Node 6: Phase B] ⚡ ${duePublications.length} publications prêtes à être expédiées.`);

    for (const pub of duePublications) {
        try {
            const topic = pub.topic || await payloadClient.getSignal(pub.topicId);
            if (!topic) {
                console.error(`🚀 [Node 6: Phase B] ❌ Topic introuvable pour la publication ${pub.id}`);
                await payloadClient.updatePublication(pub.id, { status: 'FAILED' });
                continue;
            }

            let draftData: any = {};
            try {
                draftData = JSON.parse(topic.final_draft || '{}');
            } catch (e) {
                console.error(`🚀 [Node 6: Phase B] ❌ Erreur parsing final_draft pour topic ${topic.id}`);
                await payloadClient.updatePublication(pub.id, { status: 'FAILED' });
                continue;
            }

            let isSuccess = false;

            // --- DIFFUSION DISCORD ---
            if (pub.platform === 'DISCORD') {
                const webhookUrl = process.env.DISCORD_WEBHOOK_URL;
                if (!webhookUrl) {
                    console.error("🚀 [Node 6: Phase B] ❌ DISCORD_WEBHOOK_URL absente du fichier .env");
                } else {
                    const embed = {
                        title: draftData.headline || topic.taxonomy,
                        description: draftData.body || '',
                        color: hexToDecimal('#DC2626'),
                        fields: [
                            { name: 'Niveau d\'Alerte', value: topic.taxonomy || 'INFO', inline: true },
                            { name: 'Silo Éditorial', value: topic.geo || 'Global', inline: true }
                        ],
                        footer: { text: 'Radar L\'Assez • Investigation' },
                        timestamp: new Date().toISOString()
                    };

                    const discordRes = await fetch(webhookUrl, {
                        method: 'POST',
                        headers: { 'Content-Type': 'application/json' },
                        body: JSON.stringify({ embeds: [embed] })
                    });

                    if (discordRes.ok) {
                        isSuccess = true;
                        console.log(`🚀 [Node 6: Phase B] ✅ [DISCORD] Message expédié avec succès pour le topic ${topic.id}`);
                    } else {
                        console.error(`🚀 [Node 6: Phase B] ❌ [DISCORD] Erreur HTTP ${discordRes.status}`);
                    }
                }
            }

            // --- DIFFUSION PAYLOAD CMS ---
            else if (pub.platform === 'PAYLOAD') {
                try {
                    const origin = payloadClient.apiBase.replace(/\/api\/payload$/, '');
                    const token = await payloadClient.getToken();
                    const contentPayload = draftData.body || '';
                    const topicGeo = topic.geo || 'FRANCE';
                    const topicTaxonomy = topic.taxonomy || 'INFO';
                    let niveauAlerte = 'standard';
                    if (topicTaxonomy.includes('URGENT') || topicTaxonomy.includes('FLASH')) niveauAlerte = 'flash';

                    let tagIds: string[] = [];
                    if (topic.tags) {
                        try {
                            const parsedTags = JSON.parse(topic.tags);
                            if (Array.isArray(parsedTags)) {
                                for (const tagName of parsedTags) {
                                    const cleanName = String(tagName).trim();
                                    if (cleanName) {
                                        const searchRes = await fetch(`${origin}/api/payload/tags?where[name][equals]=${encodeURIComponent(cleanName)}`);
                                        const searchData = await searchRes.json();
                                        if (searchData.docs && searchData.docs.length > 0) {
                                            tagIds.push(searchData.docs[0].id);
                                        } else {
                                            const createRes = await fetch(`${origin}/api/payload/tags`, {
                                                method: 'POST',
                                                headers: { 'Content-Type': 'application/json', 'Authorization': `JWT ${token}` },
                                                body: JSON.stringify({ name: cleanName, slug: cleanName.toLowerCase().replace(/[^a-z0-9]+/g, '-') })
                                            });
                                            const createData = await createRes.json();
                                            if (createData.doc?.id) tagIds.push(createData.doc.id);
                                        }
                                    }
                                }
                            }
                        } catch (e) { }
                    }

                    const revelationRes = await fetch(`${origin}/api/payload/revelations`, {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                            'Authorization': `JWT ${token}`
                        },
                        body: JSON.stringify({
                            titre: draftData.headline,
                            _status: 'published',
                            contenu_rapide: contentPayload,
                            contenu_rapide_html: `<p>${draftData.body.replace(/\n/g, '<br>')}</p>`,
                            niveau_alerte: niveauAlerte,
                            zone_geo: topicGeo.toLowerCase() === 'france' ? 'france' : 'international',
                            tags: tagIds.length > 0 ? tagIds : undefined
                        })
                    });

                    if (revelationRes.ok) {
                        isSuccess = true;
                        console.log(`🚀 [Node 6: Phase B] ✅ [PAYLOAD] Révélation injectée dans Payload avec succès (Topic: ${topic.id})`);
                    } else {
                        const errorText = await revelationRes.text();
                        console.error(`🚀 [Node 6: Phase B] ❌ [PAYLOAD] Échec d'injection HTTP ${revelationRes.status} :`, errorText);
                    }
                } catch(e) {
                    console.error(`🚀 [Node 6: Phase B] ❌ [PAYLOAD] Erreur lors de l'injection :`, e);
                }
            }

            // Mise à jour du statut de publication
            if (isSuccess) {
                await payloadClient.updatePublication(pub.id, {
                    status: 'PUBLISHED',
                    publishedAt: new Date()
                });

                const remaining = await payloadClient.countPendingPublications(topic.id);
                if (remaining === 0) {
                    await payloadClient.updateSignal(topic.id, {
                        status: 'PUBLISHED',
                        publishedAt: new Date()
                    });
                }
            } else {
                await payloadClient.updatePublication(pub.id, {
                    status: 'FAILED'
                });
            }

            // Pause anti-rate-limit entre les envois (2 secondes)
            await new Promise(resolve => setTimeout(resolve, 2000));
        } catch (e: any) {
            console.error(`🚀 [Node 6: Phase B] ❌ Erreur lors de la diffusion de la publication ${pub.id}:`, e.message);
            try {
                await payloadClient.updatePublication(pub.id, { status: 'FAILED' });
            } catch (err) { }
        }
    }
}

File: daemon/internal/nodes/publisher.go
package nodes

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/Flayrox/lassez.fr/daemon/internal/config"
	"github.com/Flayrox/lassez.fr/daemon/internal/store"
	"github.com/Flayrox/lassez.fr/daemon/internal/publish"
	"github.com/Flayrox/lassez.fr/daemon/internal/qoe"
)

type publisherPlatform struct {
	Name string
	Mode string
}

// RunPublisher — Node 6 : Tour de contrôle (explicite).
// Phase A (scheduling) : PENDING → QUEUED + crée Publication par plateforme activée
// Phase B (dispatch) : due PENDING → appel Channel.Publish (Discord/QOE/X/Bluesky) → PUBLISHED
// Le registry est modulaire : chaque plateforme = Channel, ajouté seulement si enableXXX=true.
func RunPublisher(client *store.Client, resolver *config.Resolver) error {
	log.Printf("\n[Node 6: Publisher] 🚀 Tour de contrôle — QOE (qoe.fi) + Discord/X/Bluesky")

	registry := buildRegistry(client, resolver)

	// ========================================================
	// PHASE A : CRÉATION DES MISSIONS DE PUBLICATION
	// ========================================================
	// Porte de modération : sans Mode Fantôme, seuls les signaux QUEUED (approuvés
	// par un humain dans le studio) sont programmés. En Mode Fantôme
	// (enableAutoApprove), les PENDING sont considérés approuvés d'office.
	autoApprove := boolParam(resolver, "publisher", "enableAutoApprove", false)
	if autoApprove {
		log.Printf("[Node 6: Phase A] 👻 Mode Fantôme : les signaux PENDING sont auto-approuvés.")
	} else {
		log.Printf("[Node 6: Phase A] ⏳ Modération active : seuls les signaux QUEUED (validation studio) sont programmés.")
	}

	pending, err := client.GetApprovableSignals(autoApprove)
	if err != nil {
		return fmt.Errorf("approvable signals: %w", err)
	}

	if len(pending) > 0 {
		log.Printf("[Node 6: Phase A] 📤 %d nouveaux articles à programmer.", len(pending))

		// Bascule atomique vers QUEUED (déjà QUEUED pour les approuvés manuels).
		ids := make([]store.ID, 0, len(pending))
		for _, t := range pending {
			ids = append(ids, t.ID)
		}
		if err := client.UpdateManySignals(ids, map[string]any{"status": "QUEUED"}); err != nil {
			log.Printf("[Node 6: Phase A] ⚠️ Bascule QUEUED partielle : %v", err)
		}

		minDelay := intParam(resolver, "publisher", "minPublishDelay", 60)
		maxDelay := intParam(resolver, "publisher", "maxPublishDelay", 120)
		if maxDelay < minDelay {
			maxDelay = minDelay
		}

		platforms := []publisherPlatform{}
		for _, name := range registry.Names() {
			mode := "DIRECT"
			switch name {
			case "DISCORD":
				mode = strParam(resolver, "publisher", "discordPublishMode", "DIRECT")
			case "X":
				mode = strParam(resolver, "publisher", "xPublishMode", "SCHEDULED")
			case "BLUESKY":
				mode = strParam(resolver, "publisher", "blueskyPublishMode", "SCHEDULED")
			case "MASTODON":
				mode = strParam(resolver, "publisher", "mastodonPublishMode", "SCHEDULED")
			}
			platforms = append(platforms, publisherPlatform{Name: name, Mode: mode})
		}

		// Matrice format → plateforme (targetsByType). La clé est l'id du format
		// (ex: "FLASH", "ALERTE") — le studio les écrit ainsi. Si la matrice est
		// absente, toutes les plateformes sont autorisées pour tous les formats.
		allowedByType := map[string]map[string]bool{}
		settings, _ := resolver.Settings()
		if raw, ok := settings["targetsByType"].(map[string]any); ok {
			for fmtID, v := range raw {
				if pm, ok := v.(map[string]any); ok {
					tgt := map[string]bool{}
					for k, val := range pm {
						b, _ := val.(bool)
						tgt[k] = b
					}
					allowedByType[fmtID] = tgt
				}
			}
		}
		// Fallback si la matrice est absente ou vide : tout est autorisé.
		resolveAllowed := func(fmtID string) map[string]bool {
			if m, ok := allowedByType[fmtID]; ok && len(m) > 0 {
				return m
			}
			// fallback : tout autorisé
			return map[string]bool{"qoe": true, "discord": true, "x": true, "bluesky": true, "mastodon": true}
		}

		// Espacement des missions par plateforme.
		lastScheduled := map[string]time.Time{}
		now := time.Now()
		for _, p := range platforms {
			if p.Mode == "SCHEDULED" {
				lastPub, err := client.GetLastScheduledPublication(p.Name)
				if err != nil {
					log.Printf("[Node 6: Phase A] ⚠️ dernière publication %s : %v", p.Name, err)
					lastScheduled[p.Name] = now
					continue
				}
				if lastPub != nil && lastPub.ScheduledAt.After(now) {
					lastScheduled[p.Name] = lastPub.ScheduledAt
				} else {
					lastScheduled[p.Name] = now
				}
			} else {
				lastScheduled[p.Name] = now
			}
		}

		var missions []store.PublicationInput
		for _, topic := range pending {
			// La matrice filtre les plateformes autorisées pour ce signal.
			// topicsWithAllowed := resolveAllowed(topic.Taxonomy)
			for _, platform := range platforms {
				// Matrice : si le format n'autorise pas cette plateforme, on saute.
				if allowed, ok := resolveAllowed(topic.Taxonomy)[strings.ToLower(platform.Name)]; ok && !allowed {
					continue
				}
				finalScheduledAt := now
				if platform.Mode == "SCHEDULED" {
					delayMinutes := minDelay + rand.Intn(maxDelay-minDelay+1) // min <= max garanti plus haut
					base := lastScheduled[platform.Name]
					finalScheduledAt = base.Add(time.Duration(delayMinutes) * time.Minute)
					lastScheduled[platform.Name] = finalScheduledAt
				}
				missions = append(missions, store.PublicationInput{
					TopicID:     topic.ID,
					Platform:    platform.Name,
					Status:      "PENDING",
					ScheduledAt: finalScheduledAt,
				})
			}
		}

		if len(missions) > 0 {
			if err := client.CreatePublications(missions); err != nil {
				log.Printf("[Node 6: Phase A] ⚠️ Création des missions : %v", err)
			} else {
				log.Printf("[Node 6: Phase A] ✅ %d missions de publication créées.", len(missions))
			}
		}
	}

	// ========================================================
	// PHASE B : EXÉCUTION DES MISSIONS PRÊTES À PARTIR
	// ========================================================
	log.Printf("[Node 6: Phase B] ⏰ Vérification des publications programmées prêtes...")

	if !boolParam(resolver, "publisher", "enableAutoPublish", true) {
		log.Printf("[Node 6: Phase B] ⏸️ Pilote automatique désactivé (enableAutoPublish=false). Diffusion ignorée.")
		return nil
	}

	due, err := client.GetDuePublications(10)
	if err != nil {
		return fmt.Errorf("due publications: %w", err)
	}
	if len(due) == 0 {
		log.Printf("[Node 6: Phase B] 📭 Aucune publication en attente pour l'instant.")
		return nil
	}
	log.Printf("[Node 6: Phase B] ⚡ %d publications prêtes à être expédiées.", len(due))

	for _, pub := range due {
		if err := dispatchPublication(client, registry, pub); err != nil {
			log.Printf("[Node 6: Phase B] ❌ Erreur lors de la diffusion de la publication %s : %v", pub.ID, err)
		}
		// Pause anti-rate-limit entre les envois.
		time.Sleep(2 * time.Second)
	}
	return nil
}

// buildRegistry instantiates the enabled channels from the config (section
// publisher, gérée par le Studio). This
// is the modular seam: every enabled platform becomes a Channel, and Phase A
// only schedules missions for registered platforms.
func buildRegistry(client *store.Client, resolver *config.Resolver) *publish.Registry {
	registry := publish.NewRegistry()

	httpClient := &http.Client{Timeout: 30 * time.Second}

	if boolParam(resolver, "publisher", "enableDiscord", true) {
		registry.Add(publish.NewDiscord(publish.DiscordConfig{
			WebhookURL: strParam(resolver, "publisher", "discordWebhookUrl", ""),
			ColorHex:   strParam(resolver, "publisher", "discordEmbedColor", "#DC2626"),
			Footer:     strParam(resolver, "publisher", "discordFooterText", "L'Assez • Investigation"),
			HTTP:       httpClient,
		}))
	}
	if boolParam(resolver, "publisher", "enableX", false) {
		registry.Add(publish.NewX(publish.XConfig{
			APIKey:           strParam(resolver, "publisher", "xApiKey", ""),
			APISecret:        strParam(resolver, "publisher", "xApiSecret", ""),
			AccessToken:      strParam(resolver, "publisher", "xAccessToken", ""),
			AccessSecret:     strParam(resolver, "publisher", "xAccessSecret", ""),
			MaxLength:        intParam(resolver, "publisher", "xMaxLength", 280),
			IncludeSourceURL: boolParam(resolver, "publisher", "includeSourceUrl", true),
			HTTP:             httpClient,
		}))
	}
	if boolParam(resolver, "publisher", "enableBluesky", false) {
		registry.Add(publish.NewBluesky(publish.BlueskyConfig{
			Identifier:       strParam(resolver, "publisher", "blueskyIdentifier", ""),
			AppPassword:      strParam(resolver, "publisher", "blueskyAppPassword", ""),
			MaxLength:        intParam(resolver, "publisher", "blueskyMaxLength", 300),
			IncludeSourceURL: boolParam(resolver, "publisher", "includeSourceUrl", true),
			HTTP:             httpClient,
		}))
	}
	if boolParam(resolver, "publisher", "enableMastodon", false) {
		registry.Add(publish.NewMastodon(publish.MastodonConfig{
			InstanceURL:      strParam(resolver, "publisher", "mastodonInstanceUrl", ""),
			AccessToken:      strParam(resolver, "publisher", "mastodonAccessToken", ""),
			MaxLength:        intParam(resolver, "publisher", "mastodonMaxLength", 500),
			IncludeSourceURL: boolParam(resolver, "publisher", "includeSourceUrl", true),
			HTTP:             httpClient,
		}))
	}
	// QOE — qoe.fi : canal de publication principal. La clé vient de
	// .secrets.yaml (qoeApiKey/qoePublicationId) via le resolver — le studio la
	// colle dans Système. Tant qu'aucune clé n'est présente : mode test (mock).
	if boolParam(resolver, "publisher", "enableQoe", true) {
		registry.Add(publish.NewQoe(publish.QoeConfig{Client: qoe.New(qoe.Config{
			BaseURL:       strParam(resolver, "publisher", "qoeBaseUrl", "https://api.qoe.fi/v1"),
			APIKey:        strParam(resolver, "publisher", "qoeApiKey", ""),
			PublicationID: strParam(resolver, "publisher", "qoePublicationId", ""),
		})}))
	}

	return registry
}

// dispatchPublication sends one due publication to its platform through the
// registry and updates the mission + signal statuses.
func dispatchPublication(client *store.Client, registry *publish.Registry, pub store.Publication) error {
	markFailed := func() {
		_ = client.UpdatePublication(pub.ID, map[string]any{"status": "FAILED"})
	}

	topic := (*store.Signal)(nil)
	if t, ok := pub.Topic(); ok {
		topic = t
	} else if topicID, ok := pub.TopicID(); ok {
		t, err := client.GetSignal(topicID)
		if err != nil {
			markFailed()
			return fmt.Errorf("récupération du signal %s : %w", topicID, err)
		}
		topic = t
	}
	if topic == nil {
		log.Printf("[Node 6: Phase B] ❌ Topic introuvable pour la publication %s", pub.ID)
		markFailed()
		return nil
	}

	var draft struct {
		Headline string `json:"headline"`
		Body     string `json:"body"`
	}
	if len(topic.FinalDraft) > 0 {
		if err := json.Unmarshal(topic.FinalDraft, &draft); err != nil {
			log.Printf("[Node 6: Phase B] ❌ Erreur parsing final_draft pour topic %s", topic.ID)
			markFailed()
			return nil
		}
	}

	channel, ok := registry.Get(pub.Platform)
	if !ok {
		log.Printf("[Node 6: Phase B] ⚠️ Plateforme %q non activée (ou inconnue).", pub.Platform)
		markFailed()
		return nil
	}

	msg := publish.Message{
		Headline: draft.Headline,
		Body:     draft.Body,
		Tags:     []string{},
		Geo:      topic.Geo,
		Taxonomy: topic.Taxonomy,
		ImageURL: topic.ImageURL,
		SignalID: string(topic.ID),
	}
	if len(topic.Tags) > 0 {
		_ = json.Unmarshal(topic.Tags, &msg.Tags)
	}
	var raw struct {
		Articles []struct {
			URL string `json:"url"`
		} `json:"articles"`
	}
	if len(topic.RawData) > 0 {
		_ = json.Unmarshal(topic.RawData, &raw)
	}
	if len(raw.Articles) > 0 {
		msg.URL = raw.Articles[0].URL
	}

	if err := channel.Publish(msg); err != nil {
		log.Printf("[Node 6: Phase B] ❌ %v", err)
		markFailed()
		return nil
	}

	now := time.Now()
	if err := client.UpdatePublication(pub.ID, map[string]any{"status": "PUBLISHED", "publishedAt": now}); err != nil {
		return err
	}
	log.Printf("[Node 6: Phase B] ✅ [%s] Message expédié pour le topic %s.", pub.Platform, topic.ID)

	if topicID, ok := pub.TopicID(); ok {
		remaining, err := client.CountPendingPublications(topicID)
		if err != nil {
			log.Printf("[Node 6: Phase B] ⚠️ CountPending %s : %v", topicID, err)
			return nil
		}
		if remaining == 0 {
			if err := client.UpdateSignal(topicID, map[string]any{"status": "PUBLISHED", "publishedAt": now}); err != nil {
				log.Printf("[Node 6: Phase B] ⚠️ Update signal %s : %v", topicID, err)
			}
		}
	}
	return nil
}

func boolParam(resolver *config.Resolver, nodeType, key string, def bool) bool {
	v := resolver.GetEffectiveParam(nodeType, key, def)
	if b, ok := v.(bool); ok {
		return b
	}
	if s, ok := v.(string); ok {
		if parsed, err := strconv.ParseBool(s); err == nil {
			return parsed
		}
	}
	return def
}

func intParam(resolver *config.Resolver, nodeType, key string, def int) int {
	v := resolver.GetEffectiveParam(nodeType, key, float64(def))
	if f, ok := v.(float64); ok {
		return int(f)
	}
	if s, ok := v.(string); ok {
		if parsed, err := strconv.Atoi(s); err == nil {
			return parsed
		}
	}
	return def
}

func strParam(resolver *config.Resolver, nodeType, key, def string) string {
	v := resolver.GetEffectiveParam(nodeType, key, def)
	if s, ok := v.(string); ok && s != "" {
		return s
	}
	return def
}
